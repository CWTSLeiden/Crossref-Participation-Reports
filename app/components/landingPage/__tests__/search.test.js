import React from 'react';
import Enzyme, { shallow, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });
import toJson from 'enzyme-to-json';
import sinon from 'sinon'

import Search from '../components/search'


let savedSearches = JSON.stringify([{name: 'saved search 1', id: 1}, {name: 'saved search 2', id: 2}])
const parsedSavedSearches = JSON.parse(savedSearches)


const localStorageMock = {
  getItem: sinon.stub().callsFake(() => savedSearches),
  setItem: sinon.stub(),
  clear: sinon.stub()
};
global.localStorage = localStorageMock;


const members = {
  'status': 'ok',
  'message-type': "members",
  "message-version": "1.0.0",
  "resultCount": 5871,
  "message": [
    {'name': 'caltech', 'id': 1},
    {'name': 'cambridge Core', 'id': 2},
    {'name': 'crossref University Press', 'id': 3},
    {'name': 'carnegie Mellon University', 'id': 4},
    {'name': 'cambridge University Press', 'id': 5},
    {'name': 'cornell University', 'id': 6}
  ]
}
global.fetch = sinon.stub().callsFake(() => Promise.resolve(members));


const history = {
  push: sinon.stub()
}
const didMount = sinon.spy(Search.prototype, 'componentDidMount')
const setState = sinon.spy(Search.prototype, 'setState')


const mountComponent = mount(<Search history={history}/>)






test('mount snapshot', () => {
  expect(didMount.callCount).toBe(1)
  expect(toJson(mountComponent)).toMatchSnapshot()
})





describe('componentDidMount', () => {

  test('should get savedSearches from localstorage', () => {
    expect(localStorageMock.getItem.callCount).toBe(1)
    expect(setState.args[0][0]).toEqual({savedSearches: parsedSavedSearches})
    expect(mountComponent.state('savedSearches')).toEqual(parsedSavedSearches)
  })

  test('should get members from api', () => {
    expect(global.fetch.callCount).toBe(1)
    expect(setState.args[1][0]).toEqual({data: members.message})
    expect(mountComponent.state('data')).toEqual(members.message)
  })
})





describe('focus input', () => {

  test('onFocus event should set focused state to true', () => {
    setState.resetHistory()

    mountComponent.find('input').prop('onFocus')()
    expect(setState.callCount).toBe(1)
    expect(mountComponent.state('focused')).toBe(true)
    mountComponent.update()
  })


  test('should show saved searches', () => {
    expect(mountComponent.state('savedSearches')).toEqual(parsedSavedSearches)

    const searchResults = mountComponent.find('.searchInputHolder').children('div')
    expect(searchResults.length).toBe(1)
    expect(searchResults.prop('style').display).toBe('initial')
    expect(mountComponent.find('Autocomplete').prop('items')).toEqual(parsedSavedSearches)
    expect(toJson(searchResults)).toMatchSnapshot()
  })


  test('onBlur event should make search results disappear', () => {
    setState.resetHistory()
    mountComponent.find('input').prop('onBlur')()
    expect(setState.callCount).toBe(1)
    expect(mountComponent.state('focused')).toBe(false)
    mountComponent.update()
    expect(mountComponent.find('.searchInputHolder').children('div').length).toBe(0)
  })


  test('if no savedSearches, results div should not be visible', () => {
    mountComponent.setState({savedSearches: undefined})
    mountComponent.update()

    mountComponent.find('input').prop('onFocus')()
    mountComponent.update()

    const searchResults = mountComponent.find('.searchInputHolder').children('div')
    expect(searchResults.length).toBe(1)
    expect(searchResults.prop('style').display).toBe('none')
  })
})





// describe('typing into search', () => {
//
//   test('...', () => {
//     mountComponent.find('input').prop('onChange')
//   })
// })
