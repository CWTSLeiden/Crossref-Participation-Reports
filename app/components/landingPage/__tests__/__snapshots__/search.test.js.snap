// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`focus input should show saved searches 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "backgroundColor": "white",
      "border": "1px black solid",
      "borderRadius": "0",
      "boxShadow": "none",
      "display": "initial",
      "left": "auto",
      "maxHeight": "207px",
      "minWidth": 0,
      "overflow": "auto",
      "padding": "13px 0 0 0",
      "position": "absolute",
      "top": "48px",
    }
  }
>
  <div
    className="searchItem"
    key="1"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    saved search 1
  </div>
  <div
    className="searchItem"
    key="2"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    saved search 2
  </div>
</div>
`;

exports[`mount snapshot 1`] = `
<Search
  history={
    Object {
      "push": [Function],
    }
  }
>
  <Autocomplete
    autoHighlight={true}
    getItemValue={[Function]}
    inputProps={Object {}}
    isItemSelectable={[Function]}
    items={
      Array [
        Object {
          "id": 1,
          "name": "saved search 1",
        },
        Object {
          "id": 2,
          "name": "saved search 2",
        },
      ]
    }
    menuStyle={
      Object {
        "backgroundColor": "white",
        "border": "1px black solid",
        "borderRadius": "0",
        "boxShadow": "none",
        "display": "initial",
        "left": "auto",
        "maxHeight": "207px",
        "overflow": "auto",
        "padding": "13px 0 0 0",
        "position": "absolute",
        "top": "48px",
      }
    }
    onChange={[Function]}
    onMenuVisibilityChange={[Function]}
    onSelect={[Function]}
    renderInput={[Function]}
    renderItem={[Function]}
    renderMenu={[Function]}
    selectOnBlur={false}
    value=""
    wrapperProps={
      Object {
        "className": "searchInputHolder",
        "style": Object {},
      }
    }
    wrapperStyle={
      Object {
        "display": "inline-block",
      }
    }
  >
    <div
      className="searchInputHolder"
      style={Object {}}
    >
      <input
        aria-autocomplete="list"
        aria-expanded={false}
        autoComplete="off"
        className="searchInput "
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="Search by member"
        role="combobox"
        value=""
      />
    </div>
  </Autocomplete>
</Search>
`;
